Deprecate redundant role vars

sevein opened the pull request:
This pull request changes the way that environment dicts are processed and interpreted.

The following is a description of the old behaviour using `archivematica_src_am_mcpclient_environment` as an example:

1. Role provides a default value for the `archivematica_src_am_mcpclient_environment` dict with preferred environment strings.
2. The user may provide a new `archivematica_src_am_mcpclient_environment` - this new dict fully replaces the default dict provided by the role which is inconvenient.
3. The user can alternatively use some role vars that map into environment strings added (e.g. `archivematica_src_search_enabled`) but only if `archivematica_src_am_mcpclient_environment` is not provided.

This pull request changes the role so:

1. Same as before - see above.
2. The user may provide a new `archivematica_src_am_mcpclient_environment` dict which is combined with the keys found in the default value.
3. The user can alternatively use other role vars that map into environment strings (e.g. `archivematica_src_search_enabled`) even when `archivematica_src_am_mcpclient_environment` is provided.

This closes #203.

sevein committed:
Deprecate redundant role vars

This closes #203.

sevein commented:
The scope of an alias node is only made available within a single YAML
file, isn't it? It's a feature only available  in the parser AFAIK. How
would that look like in this context?

I am deprecating the vars in the sense that they are not documented nor
listed as defaults but there's some code added in this PR still supporting
them to ensure backward compatibility, i.e. so your playbooks don't break.

On Tue, May 22, 2018, 20:10 Santiago Rodríguez <notifications@github.com>
wrote:

> @sevein <https://github.com/sevein> , first of all, thanks for the hard
> work doing this.
>
> What I'm not sure, is about the complexity your changes are adding to the
> role
>
> #203
> <https://github.com/artefactual-labs/ansible-archivematica-src/issues/203>
> talks about redundant vars that can be defined using different methods, and
> in our slack discussion I pointed out the problems when adding new
> environment vars listed in the install/README.md files to ansible configs.
>
> From my understanding, what this pr achieves is the ability to override
> the default environment vars defined in
> https://github.com/artefactual-labs/ansible-archivematica-src/blob/stable/1.7.x/defaults/main.yml#L64
> by adding a archivematica_src_XXX_environment> dictionary to the user
> defined conf ( per example, adding them into the host_vars//vars.yml)
>
> Another way to achieve what I think are the same results using yamI syntax
> sugar is:
>
>    - Add some meaninfull "alias" (&ansible_XXX_defaults? ) to
>    https://github.com/artefactual-labs/ansible-archivematica-src/blob/stable/1.7.x/defaults/main.yml#L64
>    and others
>    - And then, in the user defined config, an environment variable needs
>    to be added, like this:
>
>  ansible_XXX_custom_environment:
>       <<: *ansible_XXX_defaults
>       ARCHIVEMATICA_MCPCLIENT_MCPCLIENT_CAPTURE_CLIENT_SCRIPT_OUTPUT: false
>
> Found in https://stackoverflow.com/a/41823869.
>
> Also, I'm unsure about deprecating the current ansible vars in use, as
> that would break our deployments. I think of them as defaults that can be
> overriden one way or another.
>
> @hakamine <https://github.com/hakamine> , @mamedin
> <https://github.com/mamedin> , @jhsimpson <https://github.com/jhsimpson>
> , what do you think about this?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/artefactual-labs/ansible-archivematica-src/pull/204#issuecomment-391088118>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAlA-6BXvG52ZK9vv-fSvjunICyWn5z9ks5t1FSegaJpZM4UEwCJ>
> .
>

hakamine commented:
I like this! 💯  I haven't tested (yet), is there any part in the changes you think need extra attention during CR?
* If I understood the code in the PR correctly, the changes shouldn't break existing playbook configurations that use the legacy variables ( archivematica_src_search_enabled, archivematica_src_am_es_server, etc ) as their values are used by the tasks in `envs-patch-backward-compatibility.yml` 
* This PR solves what was one pain point of using dictionaries, that was the need to re-define all the keys when, for example you only needed to add a new key. With the PR now it is possible to do this.

sevein commented:
I've updated the description of the pull request hoping that it's easier to understand the changes I am suggesting. You're right @hakamine, the legacy variables are still supported.

I'm trying to discourage the use of role variables to manage application configuration but I realize that each role variable added has a different story so it's hard to reason about them, e.g. `archivematica_src_search_enabled` tries to simplify the operations that take to switch a feature toggle. Archivematica expects the same configuration attribute to be defined in three different components which wasn't a great choice but IMO it's Archivematica's responsibility to fix it. If you push that responsibility to the role then we have this unnecessary configuration mapping overhead.

I agree that there is an undesirable level of complexity being brought by this pull request. However I would say that most of it it's in `tasks/envs-patch-backward-compatibility.yml` which is the task basically undoing some recent changes in a backward-compatible manner. We don't get that for free which reminds us that we should be more cautions in the future. E.g. I approved the PR that added `archivematica_src_search_enabled` and I much regret now.

I've just rebased this branch but I'm not sure what to do with https://github.com/artefactual-labs/ansible-archivematica-src/commit/15d46f3a3fe8284211ab47726b0d74dd64167091. I don't know if we always want to patch the environment strings (e.g. `AM_GUNICORN_BIND`) after `archivematica_src_am_dashboard_nginx_bind`. I guess it makes things easier but can you think of any disadvantages?

scollazo commented on tasks/envs-patch-backward-compatibility.yml:
```diff
@@ -0,0 +1,112 @@
+
+- name: "archivematica_src_search_enabled"
+  set_fact:
+    "{{ item.name }}": "{{ item.dict|combine({item.key: archivematica_src_search_enabled}) }}"
```
This need an "always" tag, else some vars are not populated, and running other tasks fails.

scollazo reviewed (changes requested):
We need to run the new tasks always, or other tasks will fail

sevein commented:
> We need to run the new tasks always, or other tasks will fail

Done.
